From 757534e552ed01350b45f77bb55f916156b3960b Mon Sep 17 00:00:00 2001
From: Pavel Ivashchenko <pivashchenko@nfware.com>
Date: Mon, 30 Sep 2019 15:57:10 +0000
Subject: [PATCH 2/2] vtysh: fix searching commands in parent nodes

  Do not check parent command nodes in case of ambiguous and incomplete commands

Signed-off-by: Pavel Ivashchenko <pivashchenko@nfware.com>
---
 lib/command.c | 3 +++
 vtysh/vtysh.c | 2 ++
 2 files changed, 5 insertions(+)

diff --git a/lib/command.c b/lib/command.c
index 5bd1e03fc..ab256709d 100644
--- a/lib/command.c
+++ b/lib/command.c
@@ -1000,6 +1000,7 @@ int cmd_execute_command(vector vline, struct vty *vty,
 		return saved_ret;
 
 	if (ret != CMD_SUCCESS && ret != CMD_WARNING
+	    && ret != CMD_ERR_AMBIGUOUS && ret != CMD_ERR_INCOMPLETE
 	    && ret != CMD_NOT_MY_INSTANCE && ret != CMD_WARNING_CONFIG_FAILED) {
 		/* This assumes all nodes above CONFIG_NODE are childs of
 		 * CONFIG_NODE */
@@ -1011,6 +1012,7 @@ int cmd_execute_command(vector vline, struct vty *vty,
 			ret = cmd_execute_command_real(vline, FILTER_RELAXED,
 						       vty, cmd, 0);
 			if (ret == CMD_SUCCESS || ret == CMD_WARNING
+			    || ret == CMD_ERR_AMBIGUOUS || ret == CMD_ERR_INCOMPLETE
 			    || ret == CMD_NOT_MY_INSTANCE
 			    || ret == CMD_WARNING_CONFIG_FAILED)
 				return ret;
@@ -1209,6 +1211,7 @@ int command_config_read_one_line(struct vty *vty,
 	while (!(use_daemon && ret == CMD_SUCCESS_DAEMON)
 	       && !(!use_daemon && ret == CMD_ERR_NOTHING_TODO)
 	       && ret != CMD_SUCCESS && ret != CMD_WARNING
+	       && ret != CMD_ERR_AMBIGUOUS && ret != CMD_ERR_INCOMPLETE
 	       && ret != CMD_NOT_MY_INSTANCE && ret != CMD_WARNING_CONFIG_FAILED
 	       && ret != CMD_NO_LEVEL_UP)
 		ret = cmd_execute_command_real(vline, FILTER_STRICT, vty, cmd,
diff --git a/vtysh/vtysh.c b/vtysh/vtysh.c
index 74f13e1a4..be6373811 100644
--- a/vtysh/vtysh.c
+++ b/vtysh/vtysh.c
@@ -491,6 +491,7 @@ static int vtysh_execute_func(const char *line, int pager)
 	 */
 	while (ret != CMD_SUCCESS && ret != CMD_SUCCESS_DAEMON
 	       && ret != CMD_WARNING && ret != CMD_WARNING_CONFIG_FAILED
+	       && ret != CMD_ERR_AMBIGUOUS && ret != CMD_ERR_INCOMPLETE
 	       && vty->node > CONFIG_NODE) {
 		vty->node = node_parent(vty->node);
 		ret = cmd_execute(vty, line, &cmd, 1);
@@ -776,6 +777,7 @@ int vtysh_mark_file(const char *filename)
 		 */
 		while (ret != CMD_SUCCESS && ret != CMD_SUCCESS_DAEMON
 		       && ret != CMD_WARNING && ret != CMD_WARNING_CONFIG_FAILED
+		       && ret != CMD_ERR_AMBIGUOUS && ret != CMD_ERR_INCOMPLETE
 		       && vty->node > CONFIG_NODE) {
 			vty->node = node_parent(vty->node);
 			ret = cmd_execute_command_strict(vline, vty, &cmd);
-- 
2.30.2

